server/src/main/java/org/elasticsearch/common/lucene/store/BytesReferenceIndexInput.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.lucene.store;
+
+import org.apache.lucene.store.IndexInput;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+
+import java.io.EOFException;
+import java.io.IOException;
+
+public class BytesReferenceIndexInput extends IndexInput {
+
+    private final BytesReference bytesReference;
+
+    private int filePointer;
+    private StreamInput streamInput;
+
+    protected BytesReferenceIndexInput(String resourceDescription, BytesReference bytesReference) {
+        super(resourceDescription);
+        this.bytesReference = bytesReference;
+    }
+
+    @Override
+    public void close() throws IOException {}
+
+    @Override
+    public long getFilePointer() {
+        return filePointer;
+    }
+
+    private StreamInput getOrOpenStreamInput() throws IOException {
+        if (streamInput == null) {
+            streamInput = bytesReference.slice(filePointer, bytesReference.length() - filePointer).streamInput();
+        }
+        return streamInput;
+    }
+
+    @Override
+    public void seek(long longPos) throws IOException {
+        if (longPos < 0) {
+            throw new IllegalArgumentException("Seeking to negative position: " + longPos);
+        } else if (longPos > bytesReference.length()) {
+            throw new EOFException("seek past EOF");
+        }
+        var pos = (int) longPos;
+        if (streamInput != null && filePointer <= pos) {
```

Comment:
Are we never seeking backwards? seems like we need to null out the `streamInput` in that case so that `getOrOpenStreamInput` starts fresh?

Reply:
Huh. That was handled at one point, looks lost. I wonder why the tests didn't notice.

Reply:
Some kind of conservation-of-bug-count thing going on here.

https://github.com/elastic/elasticsearch/commit/7390b4dc054d5dd7b1144fac8459a4a69a7659c9#diff-1313cb767b5d3c0840a46464b34a645015f7dffb344967cf380dba0a5f54d833R65